libs/kml/color_parser.cpp

```diff
@@ -0,0 +1,122 @@
+#include "color_parser.hpp"
+
+#include "coding/hex.hpp"
+
+#include "base/string_utils.hpp"
+
+namespace kml
+{
+
+struct RGBColor
+{
+  uint8_t r, g, b;
+};
+
+std::optional<uint32_t> ParseHexColor(std::string_view c)
+{
+  if (c.empty())
+    return {};
+
+  if (c.front() == '#')
+    c.remove_prefix(1);
+  if (c.size() != 6 && c.size() != 8)
+    return {};
+
+  auto const colorBytes = FromHex(c);
+  switch (colorBytes.size())
+  {
+  case 3: return ToRGBA(colorBytes[0], colorBytes[1], colorBytes[2]);
+  case 4: return ToRGBA(colorBytes[1], colorBytes[2], colorBytes[3], colorBytes[0]);
+  default: return {};
+  }
+}
+
+// Garmin extensions spec: https://www8.garmin.com/xmlschemas/GpxExtensionsv3.xsd
+// Color mapping: https://help.locusmap.eu/topic/extend-garmin-gpx-compatibilty
+std::optional<uint32_t> ParseGarminColor(std::string_view c)
+{
+  /// @todo Unify with RGBColor instead of string.
+  static std::pair<std::string_view, std::string_view> arrColors[] = {
+      {"Black", "000000"},    {"DarkRed", "8b0000"},     {"DarkGreen", "006400"}, {"DarkYellow", "b5b820"},
+      {"DarkBlue", "00008b"}, {"DarkMagenta", "8b008b"}, {"DarkCyan", "008b8b"},  {"LightGray", "cccccc"},
+      {"DarkGray", "444444"}, {"Red", "ff0000"},         {"Green", "00ff00"},     {"Yellow", "ffff00"},
+      {"Blue", "0000ff"},     {"Magenta", "ff00ff"},     {"Cyan", "00ffff"},      {"White", "ffffff"}};
+
+  for (auto const & e : arrColors)
+    if (c == e.first)
+      return ParseHexColor(e.second);
```

Comment:
Why parsing instead of hard-coding and returning the parsed color?

Reply:
todo is above, it was taken as-is